sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+	if err != nil {
+		return err
+	}
+	if results.code == ocspStatusGood {
+		return nil
+	} else {
+		serial := issuer.SerialNumber
+		serialHex := strings.TrimSpace(certutil.GetHexFormatted(serial.Bytes(), ":"))
+		if results.code == ocspStatusRevoked {
+			return fmt.Errorf("certificate with serial number %s has been revoked", serialHex)
+		} else if conf.OcspFailureMode == FailOpenFalse {
+			return fmt.Errorf("unknown OCSP status for cert with serial number %s", strings.TrimSpace(certutil.GetHexFormatted(serial.Bytes(), ":")))
+		} else {
+			c.Logger().Warn("could not validate OCSP status for cert, but continuing in fail open mode", "serial", serialHex)
+		}
+	}
+	return nil
+}
+
+// VerifyPeerCertificate verifies all of certificate revocation status
+func (c *Client) VerifyPeerCertificate(ctx context.Context, verifiedChains [][]*x509.Certificate, conf *VerifyConfig) error {
+	for i := 0; i < len(verifiedChains); i++ {
+		// Certificate signed by Root CA. This should be one before the last in the Certificate Chain
+		numberOfNoneRootCerts := len(verifiedChains[i]) - 1
+		if !verifiedChains[i][numberOfNoneRootCerts].IsCA || string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer) != string(verifiedChains[i][numberOfNoneRootCerts].RawSubject) {
+			// Check if the last Non Root Cert is also a CA or is self signed.
+			// if the last certificate is not, add it to the list
+			rca := c.caRoot[string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer)]
+			if rca == nil {
+				return fmt.Errorf("failed to find root CA. pkix.name: %v", verifiedChains[i][numberOfNoneRootCerts].Issuer)
+			}
+			verifiedChains[i] = append(verifiedChains[i], rca)
+			numberOfNoneRootCerts++
+		}
+		results, err := c.GetAllRevocationStatus(ctx, verifiedChains[i], conf)
+		if err != nil {
+			return err
+		}
+		if r := c.canEarlyExitForOCSP(results, numberOfNoneRootCerts, conf); r != nil {
+			return r.err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) canEarlyExitForOCSP(results []*ocspStatus, chainSize int, conf *VerifyConfig) *ocspStatus {
+	msg := ""
+	if conf.OcspFailureMode == FailOpenFalse {
+		// Fail closed. any error is returned to stop connection
+		for _, r := range results {
+			if r.err != nil {
+				return r
+			}
+		}
+	} else {
+		// Fail open and all results are valid.
+		allValid := len(results) == chainSize
+		for _, r := range results {
+			if !isValidOCSPStatus(r.code) {
+				allValid = false
+				break
+			}
+		}
+		for _, r := range results {
+			if allValid && r.code == ocspStatusRevoked {
+				return r
+			}
+			if r != nil && r.code != ocspStatusGood && r.err != nil {
+				msg += "" + r.err.Error()
+			}
+		}
+	}
+	if len(msg) > 0 {
+		c.Logger().Warn(
+			"OCSP is set to fail-open, and could not retrieve OCSP based revocation checking but proceeding.", "detail", msg)
+	}
+	return nil
+}
+
+func (c *Client) ValidateWithCacheForAllCertificates(verifiedChains []*x509.Certificate) (bool, error) {
```

Comment:
Just curious why this is exported?

Reply:
Earlier version of the code called this.  Fixed.